[TRTLLM-6650][feat] Enhance beam search support with CUDA graph integration

stnie opened the pull request:
- Implemented static cache indirection buffer in model engine for beam search to allow for CUDA graph usage.
- Updated batch size calculations to account for max beam width.
- Modified dummy request creation to support variable beam widths.
- Added unit tests for beam search output shapes with CUDA graph configurations.

This change enables  beam search + CUDA graph operations in the PyTorch model engine.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Enabled support for running beam search with CUDA graphs, improving performance for advanced decoding strategies.
  * Introduced persistent cache indirection buffers to optimize beam search execution.
* **Bug Fixes**
  * Adjusted batch size and token calculations to properly account for beam width during execution and padding.
* **Tests**
  * Updated tests to verify output shapes and behavior when using beam search with CUDA graphs.
* **Refactor**
  * Enhanced attention metadata to support cache indirection handling for beam search scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

## Description

Beam Search, was not usable with CUDA Graphs. This PR enables CUDA Graph support for beam search.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

stnie committed:
feat: Enhance beam search support with CUDA graph integration

- Implemented static cache indirection buffer in model engine for beam search to allow for CUDA graph usage.
- Updated batch size calculations to account for max beam width.
- Modified dummy request creation to support variable beam widths.
- Added unit tests for beam search output shapes with CUDA graph configurations.

This change enables  beam search + CUDA graph operations in the PyTorch model engine.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #12448](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12448/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9256](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9256/) completed with status: 'SUCCESS'

stnie committed:
chore:  Adjust cache indirection passing to AttentionMetadata

- Moved the cache indirection buffer into AttentionMetadata instead of TrtllmAttentionMetadata
- Updated PyTorchModelEngine to utilize the cache indirection buffer conditionally based on the attention backend.
- Combined the beam search testcases for overlap scheduling and cuda graphs.
- Adjusted size estimation of cache indirection buffer in model_engine to correctly cover overlap scheduling

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented on tensorrt_llm/_torch/pyexecutor/model_engine.py:
```diff
@@ -735,6 +749,11 @@ def _set_up_attn_metadata(self, kv_cache_manager: KVCacheManager):
+        if self.attn_backend.Metadata is TrtllmAttentionMetadata:
+            kwargs = {"cache_indirection": self.cache_indirection_attention}
+        else:
+            kwargs = {}
```
I adjusted this. I moved the cache_indirection field to the base class of AttentionMetadata as an optional

tensorrt-cicd commented:
[PR_Github #12714](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12714/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9461](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9461/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #12726](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12726/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12821](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12821/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #12821](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/12821/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #9555](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/9555/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/9555/test-results/rerun_report.html) for details.
